tagstore/models/tag.py

```diff
@@ -1,15 +1,173 @@
 """
 Data models for tags
 """
-from typing import NamedTuple
+from typing import NamedTuple, Optional
 
+from django.core.exceptions import ValidationError as DjangoValidationError
+from django.db import models, IntegrityError
 
-class Tag(NamedTuple):
+from .common import MAX_CHAR_FIELD_LENGTH
+from .entity import EntityId, Entity
+
+
+class TagId(NamedTuple):
     """
-    A tag that can be applied to content.
+    Idenifier for a tag that can be applied to content.
     """
     # The Unique ID of the taxonomy which owns this tag
-    taxonomy_uid: int
+    taxonomy_id: int
     # The text of this tag, which also serves as its identifier.
     # Case is preserved but within a taxonomy, tags must be case-insensitively unique.
+    # Example: "good problem".
     name: str
+
+
+class Tag(models.Model):
+    """
+    A tag that can be applied to content.
+
+    A tag is a string (e.g. "difficult") and belongs to a taxonomy. The taxonomy
+    may consist of a flat set of tags or a hierarchical tree of tags.
+    """
+    # This ID should never be used - [taxonomy, name] is a compound key that
+    # should be used in all public contexts
+    _id = models.BigAutoField(primary_key=True, db_column='id')
+    # The taxonomy which owns this tag
+    taxonomy = models.ForeignKey('Taxonomy', null=False, related_name='tags')
+    # The text of this tag, which also serves as its identifier.
+    # Case is preserved but within a taxonomy, tags must be case-insensitively unique.
+    # Example: "good problem".
+    name = models.CharField(max_length=MAX_CHAR_FIELD_LENGTH, db_column='tag')
+    # Materialized path. Always ends with ":".
+    # A simple tag like "good problem" would have a path of "good problem:"
+    # A tag like "mammal" that is a child of "animal" would have a path of
+    # "animal:mammal:". Tags are not allowed to contain the ":" character
+    # so no escaping is necessary.
+    path = models.CharField(max_length=MAX_CHAR_FIELD_LENGTH, db_index=True)
+
+    PATH_SEP = ':'  # Character used to separate tags
```

Comment:
Possibly move this constant to common.py.

Reply:
It's only used in this file though?